packages/@uppy/companion-client/src/RequestClient.js

```diff
@@ -94,92 +89,80 @@ export default class RequestClient {
     return `${this.hostname}/${url}`
   }
 
-  #errorHandler (method, path) {
-    return (err) => {
-      if (!err?.isAuthError) {
-        // eslint-disable-next-line no-param-reassign
-        err = new ErrorWithCause(`Could not ${method} ${this.#getUrl(path)}`, { cause: err })
+  /*
+    Preflight was added to avoid breaking change between older Companion-client versions and
+    newer Companion versions and vice-versa. Usually the break will manifest via CORS errors because a
+    version of companion-client could be sending certain headers to a version of Companion server that
+    does not support those headers. In which case, the default preflight would lead to CORS.
+    So to avoid those errors, we do preflight ourselves, to see what headers the Companion server
+    we are communicating with allows. And based on that, companion-client knows what headers to
+    send and what headers to not send.
+
+    The preflight only happens once throughout the life-cycle of a certain
+    Companion-client <-> Companion-server pair (allowedHeadersCache).
+    Subsequent requests use the cached result of the preflight.
+  */
+  async preflight (path) {
+    const allowedHeadersCached = allowedHeadersCache.get(this.hostname)
+    if (allowedHeadersCached != null) return allowedHeadersCached
+
+    const promise = (async () => {
+      try {
+        const response = await fetch(this.#getUrl(path), { method: 'OPTIONS' })
+
+        const header = response.headers.get('access-control-allow-headers')
+        if (header == null) return undefined
+
+        this.uppy.log(`[CompanionClient] adding allowed preflight headers to companion cache: ${this.hostname} ${header}`)
+
+        const allowedHeaders = header.split(',').map((headerName) => headerName.trim().toLowerCase())
+        allowedHeadersCache.set(this.hostname, allowedHeaders)
+        return allowedHeaders
+      } catch (err) {
+        this.uppy.log(`[CompanionClient] unable to make preflight request ${err}`, 'warning')
+        return undefined
```

Comment:
We probably need to cache the error case as well.
```suggestion
        allowedHeadersCache.set(this.hostname, fallbackAllowedHeaders)
        return undefined
```